### PR TITLE
Add normalizer sample

### LINKED ISSUE #5: Create a sample for Normalizer
Create a sample for [Normalizer](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Normalizer.html).

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
Fixes #5

### CODE REVIEW

File: normalizer/normalizer.bal [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+function validateXmlRequest(xml request) returns boolean =>
+    (request/<subject>).data() == "Printer offline"
+        && (request/<comment>).data() == "My printer is offline. Restarting doesn’t help.";
+
+function validateJsonRequest(json request) returns boolean => {
+        "subject": "Printer offline",
+        "comment": "My printer is offline. Restarting doesn’t help."
+    } == request;
```
Comment (ushirask) [inline normalizer/normalizer.bal]: I don't think we should be validating for the hard coded request here

File: normalizer/normalizer.bal [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                    <id>${result.ticket.id}</id>
+                    <subject>${result.ticket.subject}</subject>
+                    <url>${result.ticket.url}</url>
+                    </response>`;
+            } else {
+                return {
+                    id: result.ticket.id,
+                    subject: result.ticket.subject,
+                    url: result.ticket.url
+                };
```
Comment (ushirask) [inline normalizer/normalizer.bal]: Do we need to have different response formats? IMO since it's not part of the Normalizer pattern it would be more clear to return a common response irrespective of the request format

File: normalizer/normalizer.bal [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            }
+        }
+        return error ("Invalid request");
```
Comment (ushirask) [inline normalizer/normalizer.bal]:
```suggestion
        return error("Invalid request");
```

File: normalizer/normalizer.bal (hunk 2 above)
Comment (LakshanWeerasinghe) [reply normalizer/normalizer.bal]: @manuranga wdyt?

File: normalizer/normalizer.bal [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        json zendeskRequest = {};
+        if request is json && validateJsonRequest(request) {
+            zendeskRequest = {
+                ticket: {
+                    subject: check request.subject, 
+                    comment: {
+                        body: check request.comment
+                        }
+                    }
+                };
```
Comment (DimuthuMadushan) [inline normalizer/normalizer.bal]:
```suggestion
            zendeskRequest = {
                ticket: {
                    subject: check request.subject, 
                    comment: {
                        body: check request.comment
                    }
                }
            };
```

File: normalizer/normalizer.bal [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        if request is xml && validateXmlRequest(request) {
+            zendeskRequest = {
+                ticket: {
+                    subject: (request/<subject>).data(), 
+                    comment: {
+                        body: (request/<comment>).data()
+                        }
+                    }
+                };
```
Comment (DimuthuMadushan) [inline normalizer/normalizer.bal]:
```suggestion
            zendeskRequest = {
                ticket: {
                    subject: (request/<subject>).data(), 
                    comment: {
                        body: (request/<comment>).data()
                    }
                }
            };
```

File: normalizer/normalizer.bal [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            http:Client zendeskClient = check new ("http://api.zendesk.com.balmock.io");
+            ZendeskResponse result = check zendeskClient->/api/v2/tickets.post(zendeskRequest);
+            if request is xml {
+                return xml `
+                    <response>
+                    <id>${result.ticket.id}</id>
+                    <subject>${result.ticket.subject}</subject>
+                    <url>${result.ticket.url}</url>
+                    </response>`;
+            } else {
```
Comment (DimuthuMadushan) [inline normalizer/normalizer.bal]: We can remove `else` block here.

File: normalizer/normalizer.bal (hunk 2 above)
Comment (manuranga) [reply normalizer/normalizer.bal]: +1 for @ushirask's suggestion.

File: normalizer/normalizer.bal (hunk 1 above)
Comment (manuranga) [reply normalizer/normalizer.bal]: +1, we can't validated against hard-coded value.

File: normalizer/tests/case-01.http [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+###
+
+POST http://localhost:8080/api/v1/ticket
+Content-Type: application/xml
+
+<ticket>
+    <subject>Printer offline</subject>
+    <comment>My printer is offline. Restarting doesn’t help.</comment>
+</ticket>
```
Comment (manuranga) [inline normalizer/tests/case-01.http]: shouldn't this go to a `case-02.http`?

Comment (LakshanWeerasinghe) [reply normalizer/tests/case-01.http]: Yes

File: normalizer/normalizer.bal [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        if zendeskRequest != {} {
+            http:Client zendeskClient = check new ("http://api.zendesk.com.balmock.io");
+            ZendeskResponse result = check zendeskClient->/api/v2/tickets.post(zendeskRequest);
+            return {
+                id: result.ticket.id,
+                subject: result.ticket.subject,
+                url: result.ticket.url
+            };
+        }
+        return error("Invalid request");
```
Comment (manuranga) [inline normalizer/normalizer.bal:44]: This is not possible. Refactor code to remove this.

Comment (manuranga) [inline normalizer/normalizer.bal]: This is not possible. Refactor code to remove this.

**Review by manuranga — CHANGES_REQUESTED** [review]

File: normalizer/normalizer.bal [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    record {|
+        string url;
+        int id;
+        string subject;
+    |} ticket;
+};
+
+service /api/v1 on new http:Listener(8080) {
+    
+    resource function post ticket(@http:Payload json|xml request) returns xml|json|error {
```
Comment (anupama-pathirage) [inline normalizer/normalizer.bal]: We can omit `@http:Payload` part.

Comment (LakshanWeerasinghe) [reply normalizer/normalizer.bal]: Removing that will gives a compile time error.  `invalid union type for default payload param: request. Use basic structured anydata types`

File: normalizer/normalizer.bal [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        if request is xml {
+            zendeskRequest = {
+                ticket: {
+                    subject: (request/<subject>).data(), 
+                    comment: {
+                        body: (request/<comment>).data()
+                    }
+                }
+            };
+        }
```
Comment (manuranga) [inline normalizer/normalizer.bal]: Let's move this to a function called `normalize`

File: normalizer/normalizer.bal [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        if zendeskRequest != {} {
```
Comment (manuranga) [inline normalizer/normalizer.bal]: this can never happen.

File: normalizer/normalizer.bal [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            http:Client zendeskClient = check new ("http://api.zendesk.com.balmock.io");
+            ZendeskResponse result = check zendeskClient->/api/v2/tickets.post(zendeskRequest);
+            return {
+                id: result.ticket.id,
+                subject: result.ticket.subject,
+                url: result.ticket.url
+            };
```
Comment (manuranga) [inline normalizer/normalizer.bal]: Let's return the URL or id to simplify.

Comment (LakshanWeerasinghe) [reply normalizer/normalizer.bal]: returned a tuple of [int, string, string].

Comment (LakshanWeerasinghe) [reply normalizer/normalizer.bal]: I guess this is simple and required only one line to write the code.

File: normalizer/normalizer.bal [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        string subject;
+    |} ticket;
+};
+
+http:Client zendeskClient = check new ("http://api.zendesk.com.balmock.io");
+
+service /api/v1 on new http:Listener(8080) {
+    
+    resource function post ticket(@http:Payload json|xml request) returns [int, string, string]|error {
+        json zendeskRequest;
```
Comment (manuranga) [inline normalizer/normalizer.bal]: Let's call this `normalizedRequest`

File: normalizer/normalizer.bal [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        if request is json {
+            zendeskRequest = normalize(check request.subject, check request.comment);
+        } else {
+            zendeskRequest = normalize((request/<subject>).data(), (request/<comment>).data());
+        }
+        ZendeskResponse zendeskResponse = check zendeskClient->/api/v2/tickets.post(zendeskRequest);
+        return [zendeskResponse.ticket.id, zendeskResponse.ticket.subject, zendeskResponse.ticket.url];
```
Comment (manuranga) [inline normalizer/normalizer.bal]: Isn't conceptually this same as `return check zendeskClient->/api/v2/tickets.post(zendeskRequest);`? now it returns a json array instead json map. But that is a superficial change. It's still a json.

File: mock-http/Dependencies.toml [hunk 15]
```diff
@@ -0,0 +1,289 @@
+# AUTO-GENERATED FILE. DO NOT MODIFY.
```
Comment (manuranga) [inline mock-http/Dependencies.toml]: This shouldn't be part of this pr.

**Review by manuranga — CHANGES_REQUESTED** [review]

File: normalizer/normalizer.bal (hunk 14 above)
Comment (LakshanWeerasinghe) [reply normalizer/normalizer.bal]: Updated to return the url of the ticket.

File: normalizer/normalizer.bal [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+            normalizedRequest = normalize((request/<subject>).data(), (request/<comment>).data());
+        }
+        ZendeskResponse zendeskResponse = check zendeskClient->/api/v2/tickets.post(normalizedRequest);
+        return zendeskResponse.ticket.url;
+    }       
+}
+
+function normalize(string subject, string comment) returns json {
+    return {
+        "ticket": {
```
Comment (manuranga) [inline normalizer/normalizer.bal]: let's get rid of the quote.

**Review by manuranga — APPROVED** [review]
